// @microsoft/teamsfx-core/build/component/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityOutputs = exports.AzureSqlOutputs = exports.ErrorConstants = exports.TelemetryConstants = exports.BicepConstants = exports.ActionTypeShell = exports.ActionTypeGroup = exports.ActionTypeCall = exports.ActionTypeFunction = exports.ActionNames = exports.ComponentStateKeys = exports.ComponentNames = void 0;
exports.ComponentNames = {
    TeamsTab: "teams-tab",
    TeamsBot: "teams-bot",
    AppManifest: "app-manifest",
    AadApp: "aad-app",
    AzureWebApp: "azure-web-app",
    AzureStorage: "azure-storage",
    BotService: "bot-service",
    SPFx: "spfx",
    Identity: "identity",
    APIM: "apim",
    KeyVault: "key-vault",
    AzureSQL: "azure-sql",
    TabCode: "tab-code",
    BotCode: "bot-code",
    Function: "azure-function",
    SimpleAuth: "simple-auth",
};
exports.ComponentStateKeys = {
    [exports.ComponentNames.AzureSQL]: "azure-sql",
    [exports.ComponentNames.Identity]: "identity",
    [exports.ComponentNames.AadApp]: "aad",
    [exports.ComponentNames.KeyVault]: "key-vault",
};
exports.ActionNames = {
    provision: "provision",
    configure: "configure",
    generateBicep: "generateBicep",
};
exports.ActionTypeFunction = "function";
exports.ActionTypeCall = "call";
exports.ActionTypeGroup = "group";
exports.ActionTypeShell = "shell";
exports.BicepConstants = {
    writeFile: "1",
};
exports.TelemetryConstants = {
    eventPrefix: "-start",
    properties: {
        component: "component",
        success: "success",
        errorCode: "error-code",
        errorType: "error-type",
        errorMessage: "error-message",
    },
    values: {
        yes: "yes",
        no: "no",
        userError: "user",
        systemError: "system",
    },
};
exports.ErrorConstants = {
    unhandledError: "UnhandledError",
    unhandledErrorMessage: "Unhandled Error",
};
exports.AzureSqlOutputs = {
    sqlResourceId: {
        key: "resourceId",
        bicepVariable: "provisionOutputs.azureSqlOutput.value.resourceId",
    },
    sqlEndpoint: {
        key: "endpoint",
        bicepVariable: "provisionOutputs.azureSqlOutput.value.sqlEndpoint",
    },
    sqlDatabaseName: {
        key: "databaseName",
        bicepVariable: "provisionOutputs.azureSqlOutput.value.sqlDatabaseName",
    },
};
exports.IdentityOutputs = {
    identityResourceId: {
        key: "resourceId",
        bicepVariable: "provisionOutputs.identityOutput.value.identityResourceId",
    },
    identityName: {
        key: "identityName",
        bicepVariable: "provisionOutputs.identityOutput.value.identityName",
    },
    identityClientId: {
        key: "identityClientId",
        bicepVariable: "provisionOutputs.identityOutput.value.identityClientId",
    },
};
