// @microsoft/teamsfx-core/build/component/resource/azureAppService/azureFunction.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureFunctionResource = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const typedi_1 = require("typedi");
const azureAppService_1 = require("./azureAppService");
let AzureFunctionResource = class AzureFunctionResource extends azureAppService_1.AzureAppService {
    constructor() {
        super(...arguments);
        this.name = "azure-function";
        this.alias = "FT";
        this.displayName = "Azure Functions";
        this.bicepModuleName = "azureFunction";
        this.outputs = {
            functionAppResourceId: {
                key: "functionAppResourceId",
                bicepVariable: "provisionOutputs.azureFunction{{componentName}}Output.value.resourceId",
            },
            functionEndpoint: {
                key: "functionEndpoint",
                bicepVariable: "azureFunction{{componentName}}Provision.outputs.endpoint",
            },
        };
        this.finalOutputKeys = ["resourceId", "endpoint"];
    }
    configure(context, inputs) {
        const action = {
            name: "azure-function.configure",
            type: "function",
            plan: (context, inputs) => {
                return teamsfx_api_1.ok([
                    {
                        type: "service",
                        name: "azure",
                        remarks: "config azure function",
                    },
                ]);
            },
            execute: async (context, inputs) => {
                // Configure APIM
                return teamsfx_api_1.ok([
                    {
                        type: "service",
                        name: "azure",
                        remarks: "config azure function",
                    },
                ]);
            },
        };
        return teamsfx_api_1.ok(action);
    }
};
AzureFunctionResource = tslib_1.__decorate([
    typedi_1.Service("azure-function")
], AzureFunctionResource);
exports.AzureFunctionResource = AzureFunctionResource;
