// @microsoft/teamsfx-core/build/component/resource/azureSql/azureSql.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureSqlResource = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
require("reflect-metadata");
const typedi_1 = require("typedi");
const constants_1 = require("../../constants");
const configure_1 = require("./actions/configure");
const generateBicep_1 = require("./actions/generateBicep");
const provision_1 = require("./actions/provision");
let AzureSqlResource = class AzureSqlResource {
    constructor() {
        this.name = constants_1.ComponentNames.AzureSQL;
        this.outputs = constants_1.AzureSqlOutputs;
        this.finalOutputKeys = ["sqlResourceId", "endpoint", "databaseName"];
    }
    generateBicep(context, inputs) {
        return teamsfx_api_1.ok(generateBicep_1.GetActionGenerateBicep());
    }
    provision(context, inputs) {
        return teamsfx_api_1.ok(provision_1.ProvisionActionImplement.get());
    }
    configure(context, inputs) {
        return teamsfx_api_1.ok(configure_1.ConfigureActionImplement.get());
    }
};
AzureSqlResource = tslib_1.__decorate([
    typedi_1.Service("azure-sql")
], AzureSqlResource);
exports.AzureSqlResource = AzureSqlResource;
