// @microsoft/teamsfx-core/build/plugins/resource/appstudio/appStudio.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppStudioClient = exports.getAppStudioEndpoint = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const errors_1 = require("./errors");
const results_1 = require("./results");
const constants_1 = require("./constants");
const utils_1 = require("./utils/utils");
const telemetry_1 = require("./utils/telemetry");
function getAppStudioEndpoint() {
    if (process.env.APP_STUDIO_ENV && process.env.APP_STUDIO_ENV === "int") {
        return "https://dev-int.teams.microsoft.com";
    }
    else {
        return "https://dev.teams.microsoft.com";
    }
}
exports.getAppStudioEndpoint = getAppStudioEndpoint;
// eslint-disable-next-line @typescript-eslint/no-namespace
var AppStudioClient;
(function (AppStudioClient) {
    const baseUrl = getAppStudioEndpoint();
    /**
     * Creates a new axios instance to call app studio to prevent setting the accessToken on global instance.
     * @param {string}  appStudioToken
     * @returns {AxiosInstance}
     */
    function createRequesterWithToken(appStudioToken) {
        const instance = axios_1.default.create({
            baseURL: baseUrl,
        });
        instance.defaults.headers.common["Authorization"] = `Bearer ${appStudioToken}`;
        instance.defaults.headers.common["Client-Source"] = "teamstoolkit";
        instance.interceptors.request.use(function (config) {
            config.params = Object.assign({ teamstoolkit: true }, config.params);
            return config;
        });
        return instance;
    }
    function wrapException(e, apiName) {
        var _a, _b, _c, _d, _e;
        const correlationId = (_a = e.response) === null || _a === void 0 ? void 0 : _a.headers[constants_1.Constants.CORRELATION_ID];
        const requestPath = ((_b = e.request) === null || _b === void 0 ? void 0 : _b.path) ? `${e.request.method} ${e.request.path}` : "";
        const extraData = ((_c = e.response) === null || _c === void 0 ? void 0 : _c.data) ? `data: ${JSON.stringify(e.response.data)}` : "";
        const error = results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.DeveloperPortalAPIFailedError.name, errors_1.AppStudioError.DeveloperPortalAPIFailedError.message(e, correlationId, requestPath, apiName, extraData));
        telemetry_1.TelemetryUtils.sendErrorEvent(telemetry_1.TelemetryEventName.appStudioApi, error, {
            method: (_d = e.request) === null || _d === void 0 ? void 0 : _d.method,
            "status-code": (_e = e === null || e === void 0 ? void 0 : e.response) === null || _e === void 0 ? void 0 : _e.status,
            url: `<${apiName}-url>`,
        });
        return error;
    }
    /**
     * Creates an app registration in app studio with the given archived file and returns the app definition.
     * @param {Buffer}  file - Zip file with manifest.json and two icons
     * @param {string}  appStudioToken
     * @param {LogProvider} logProvider
     * @returns {Promise<AppDefinition>}
     */
    async function createApp(file, appStudioToken, logProvider) {
        try {
            const requester = createRequesterWithToken(appStudioToken);
            const response = await utils_1.RetryHandler.Retry(() => requester.post(`/api/appdefinitions/v2/import`, file, {
                headers: { "Content-Type": "application/zip" },
            }));
            if (response && response.data) {
                const app = response.data;
                await (logProvider === null || logProvider === void 0 ? void 0 : logProvider.debug(`recieved data from app studio ${JSON.stringify(app)}`));
                return app;
            }
            else {
                throw new Error(`Cannot create teams app`);
            }
        }
        catch (e) {
            const error = wrapException(e, "create-app");
            throw error;
        }
    }
    AppStudioClient.createApp = createApp;
    async function uploadIcon(teamsAppId, appStudioToken, colorIconContent, outlineIconContent, logProvider) {
        await (logProvider === null || logProvider === void 0 ? void 0 : logProvider.info(`Uploading icon for teams ${teamsAppId}`));
        const requester = createRequesterWithToken(appStudioToken);
        try {
            const colorIcon = {
                name: "color",
                type: "color",
                base64String: colorIconContent,
            };
            const outlineIcon = {
                name: "outline",
                type: "outline",
                base64String: outlineIconContent,
            };
            const colorIconResult = requester.post(`/api/appdefinitions/${teamsAppId}/image`, colorIcon);
            const outlineIconResult = requester.post(`/api/appdefinitions/${teamsAppId}/image`, outlineIcon);
            const results = await Promise.all([colorIconResult, outlineIconResult]);
            await (logProvider === null || logProvider === void 0 ? void 0 : logProvider.info(`successfully uploaded two icons`));
            return { colorIconUrl: results[0].data, outlineIconUrl: results[1].data };
        }
        catch (e) {
            const error = wrapException(e, "update-icon");
            throw error;
        }
    }
    async function getApp(teamsAppId, appStudioToken, logProvider) {
        const requester = createRequesterWithToken(appStudioToken);
        try {
            const response = await utils_1.RetryHandler.Retry(() => requester.get(`/api/appdefinitions/${teamsAppId}`));
            if (response && response.data) {
                const app = response.data;
                if (app && app.teamsAppId && app.teamsAppId === teamsAppId) {
                    return app;
                }
                else {
                    await (logProvider === null || logProvider === void 0 ? void 0 : logProvider.error(`teamsAppId mismatch. Input: ${teamsAppId}. Got: ${app.teamsAppId}`));
                }
            }
        }
        catch (e) {
            const error = wrapException(e, "get-app");
            throw error;
        }
        throw new Error(`Cannot get the app definition with app ID ${teamsAppId}`);
    }
    AppStudioClient.getApp = getApp;
    /**
     * Updates an existing app if it exists with the configuration given.  Returns whether or not it was successful.
     * @param {string}  teamsAppId
     * @param {IAppDefinition} appDefinition
     * @param {string}  appStudioToken
     * @param {LogProvider} logProvider
     * @param {string} colorIconContent - base64 encoded
     * @param {string} outlineIconContent - base64 encoded
     * @returns {Promise<IAppDefinition>}
     */
    async function updateApp(teamsAppId, appDefinition, appStudioToken, logProvider, colorIconContent, outlineIconContent) {
        // Get userlist from existing app
        const existingAppDefinition = await getApp(teamsAppId, appStudioToken, logProvider);
        const userlist = existingAppDefinition.userList;
        appDefinition.userList = userlist;
        let result;
        if (colorIconContent && outlineIconContent) {
            result = await uploadIcon(teamsAppId, appStudioToken, colorIconContent, outlineIconContent, logProvider);
            if (!result) {
                await (logProvider === null || logProvider === void 0 ? void 0 : logProvider.error(`failed to upload color icon for: ${teamsAppId}`));
                throw new Error(`failed to upload icons for ${teamsAppId}`);
            }
            appDefinition.colorIcon = result.colorIconUrl;
            appDefinition.outlineIcon = result.outlineIconUrl;
        }
        const requester = createRequesterWithToken(appStudioToken);
        try {
            const response = await requester.post(`/api/appdefinitions/${teamsAppId}/override`, appDefinition);
            if (response && response.data) {
                const app = response.data;
                return app;
            }
            else {
                throw new Error(`Cannot update teams app ${teamsAppId}, response: ${JSON.stringify(response)}`);
            }
        }
        catch (e) {
            const error = wrapException(e, "update-app");
            throw error;
        }
    }
    AppStudioClient.updateApp = updateApp;
    /**
     * Publish Teams app to Teams App Catalog
     * @param teamsAppId
     * @param file
     * @param appStudioToken
     * @returns
     */
    async function publishTeamsApp(teamsAppId, file, appStudioToken) {
        var _a, _b, _c, _d, _e, _f;
        try {
            const requester = createRequesterWithToken(appStudioToken);
            const response = await utils_1.RetryHandler.Retry(() => requester.post("/api/publishing", file, {
                headers: { "Content-Type": "application/zip" },
            }));
            if (response && response.data) {
                const requestPath = `${(_a = response.request) === null || _a === void 0 ? void 0 : _a.method} ${(_b = response.request) === null || _b === void 0 ? void 0 : _b.path}`;
                if (response.data.error) {
                    // To avoid App Studio BadGateway error
                    // The app is actually published to app catalog.
                    if (response.data.error.code === "BadGateway") {
                        const appDefinition = await getAppByTeamsAppId(teamsAppId, appStudioToken);
                        if (appDefinition) {
                            return appDefinition.teamsAppId;
                        }
                    }
                    throw results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, errors_1.AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, requestPath), `code: ${response.data.error.code}, message: ${response.data.error.message}`);
                }
                else {
                    return response.data.id;
                }
            }
            else {
                throw results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, errors_1.AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, "POST /api/publishing"));
            }
        }
        catch (e) {
            if (e instanceof teamsfx_api_1.SystemError) {
                throw e;
            }
            else {
                const correlationId = (_c = e.response) === null || _c === void 0 ? void 0 : _c.headers[constants_1.Constants.CORRELATION_ID];
                const requestPath = `${(_e = (_d = e.response) === null || _d === void 0 ? void 0 : _d.request) === null || _e === void 0 ? void 0 : _e.method} ${(_f = e.response.request) === null || _f === void 0 ? void 0 : _f.path}`;
                throw results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, errors_1.AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, requestPath, correlationId), e);
            }
        }
    }
    AppStudioClient.publishTeamsApp = publishTeamsApp;
    /**
     * Update existed publish request
     * @param teamsAppId
     * @param file
     * @param appStudioToken
     * @returns
     */
    async function publishTeamsAppUpdate(teamsAppId, file, appStudioToken) {
        var _a, _b, _c, _d, _e, _f, _g;
        try {
            // Get App Definition from Teams App Catalog
            const appDefinition = await getAppByTeamsAppId(teamsAppId, appStudioToken);
            const requester = createRequesterWithToken(appStudioToken);
            let response = null;
            if (appDefinition) {
                // update the existing app
                response = await requester.post(`/api/publishing/${appDefinition.teamsAppId}/appdefinitions`, file, { headers: { "Content-Type": "application/zip" } });
            }
            else {
                throw results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, errors_1.AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, `GET /api/publishing/${teamsAppId}`));
            }
            const requestPath = `${(_a = response.request) === null || _a === void 0 ? void 0 : _a.method} ${(_b = response.request) === null || _b === void 0 ? void 0 : _b.path}`;
            if (response && response.data) {
                if (response.data.error || response.data.errorMessage) {
                    throw results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, errors_1.AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, requestPath), ((_c = response.data.error) === null || _c === void 0 ? void 0 : _c.message) || response.data.errorMessage);
                }
                else {
                    return response.data.teamsAppId;
                }
            }
            else {
                throw results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, errors_1.AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, requestPath));
            }
        }
        catch (error) {
            if (error instanceof teamsfx_api_1.SystemError) {
                throw error;
            }
            else {
                const correlationId = (_d = error.response) === null || _d === void 0 ? void 0 : _d.headers[constants_1.Constants.CORRELATION_ID];
                const requestPath = `${(_f = (_e = error.response) === null || _e === void 0 ? void 0 : _e.request) === null || _f === void 0 ? void 0 : _f.method} ${(_g = error.response.request) === null || _g === void 0 ? void 0 : _g.path}`;
                throw results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppPublishFailedError.name, errors_1.AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, requestPath, correlationId), error);
            }
        }
    }
    AppStudioClient.publishTeamsAppUpdate = publishTeamsAppUpdate;
    async function getAppByTeamsAppId(teamsAppId, appStudioToken) {
        const requester = createRequesterWithToken(appStudioToken);
        const response = await requester.get(`/api/publishing/${teamsAppId}`);
        if (response && response.data && response.data.value && response.data.value.length > 0) {
            const appdefinitions = response.data.value[0].appDefinitions.map((item) => {
                return {
                    lastModifiedDateTime: item.lastModifiedDateTime
                        ? new Date(item.lastModifiedDateTime)
                        : null,
                    publishingState: item.publishingState,
                    teamsAppId: item.teamsAppId,
                    displayName: item.displayName,
                };
            });
            return appdefinitions[appdefinitions.length - 1];
        }
        else {
            return undefined;
        }
    }
    AppStudioClient.getAppByTeamsAppId = getAppByTeamsAppId;
    async function getUserList(teamsAppId, appStudioToken) {
        let app;
        try {
            app = await getApp(teamsAppId, appStudioToken);
        }
        catch (error) {
            throw error;
        }
        return app.userList;
    }
    AppStudioClient.getUserList = getUserList;
    async function checkPermission(teamsAppId, appStudioToken, userObjectId) {
        let userList;
        try {
            userList = await getUserList(teamsAppId, appStudioToken);
        }
        catch (error) {
            return constants_1.Constants.PERMISSIONS.noPermission;
        }
        const findUser = userList === null || userList === void 0 ? void 0 : userList.find((user) => user.aadId === userObjectId);
        if (!findUser) {
            return constants_1.Constants.PERMISSIONS.noPermission;
        }
        if (findUser.isAdministrator) {
            return constants_1.Constants.PERMISSIONS.admin;
        }
        else {
            return constants_1.Constants.PERMISSIONS.operative;
        }
    }
    AppStudioClient.checkPermission = checkPermission;
    async function grantPermission(teamsAppId, appStudioToken, newUser) {
        var _a, _b;
        let app;
        try {
            app = await getApp(teamsAppId, appStudioToken);
        }
        catch (error) {
            throw error;
        }
        if (checkUser(app, newUser)) {
            return;
        }
        const findUser = (_a = app.userList) === null || _a === void 0 ? void 0 : _a.findIndex((user) => user["aadId"] === newUser.aadId);
        if (findUser && findUser >= 0) {
            return;
        }
        (_b = app.userList) === null || _b === void 0 ? void 0 : _b.push(newUser);
        const requester = createRequesterWithToken(appStudioToken);
        const response = await requester.post(`/api/appdefinitions/${teamsAppId}/owner`, app);
        if (!response || !response.data || !checkUser(response.data, newUser)) {
            throw new Error(constants_1.ErrorMessages.GrantPermissionFailed);
        }
    }
    AppStudioClient.grantPermission = grantPermission;
    function checkUser(app, newUser) {
        var _a;
        const findUser = (_a = app.userList) === null || _a === void 0 ? void 0 : _a.findIndex((user) => user["aadId"] === newUser.aadId);
        if (findUser != undefined && findUser >= 0) {
            return true;
        }
        else {
            return false;
        }
    }
})(AppStudioClient = exports.AppStudioClient || (exports.AppStudioClient = {}));
